components/iosnoop/src/biosnoop.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::{IOStats, IOType};
+use bcc::{table::Table, Kprobe, BPF};
+use std::collections::HashMap;
+use std::ptr;
+
+static mut BPF_TABLE: Option<(BPF, Table, Table)> = None;
+
+use crossbeam_utils::CachePadded;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+static IDX_COUNTER: AtomicUsize = AtomicUsize::new(0);
+// For simplicity, just open large enough array. TODO: make it to be Vec
+static mut IO_TYPE_ARRAY: [CachePadded<IOType>; 100] = [CachePadded::new(IOType::Other); 100];
+
+thread_local! {
+    static IDX: usize = unsafe {
+        let idx = IDX_COUNTER.fetch_add(1, Ordering::SeqCst);
+        if idx == 100 {
+            panic!("exceed maximum thread count");
+        }
+        if let Some((_, _, t)) = BPF_TABLE.as_mut() {
+            let tid = nix::unistd::gettid().as_raw() as u32;
+            let ptr : *const *const _ = &IO_TYPE_ARRAY.as_ptr().add(idx);
+            println!("tid {} ptr is {:?}", tid, IO_TYPE_ARRAY.as_ptr().add(idx));
+            *IO_TYPE_ARRAY[idx] = IOType::Compaction;
+            println!("{:?}", ptr::read(IO_TYPE_ARRAY.as_ptr().add(idx) as *const IOType));
+            let io_type_ptr =
+                std::slice::from_raw_parts_mut(ptr as *mut u8, std::mem::size_of::<*const IOType>());
+            t.set(&mut tid.to_ne_bytes(), io_type_ptr).unwrap();
+        }
```

Comment:
Should we print out current thread isn't tracked?

Reply:
isn't tracked?

Reply:
you mean when BPF_TABLE is none? An error message will be printed when calling `init_io_snooper`.